### PR TITLE
gh-119004: fix a crash in equality testing between `OrderedDict`

### PR DESCRIPTION
**Pull request by picnixz**
This patch is based on https://github.com/python/cpython/issues/119004#issuecomment-2108272841 but some bits needed to be added here and there.

* Issue: gh-119004

----
📚 Documentation preview 📚: https://cpython-previews--121329.org.readthedocs.build/

### CODE REVIEW

**Comment by rhettinger** [comment]
I don't see the need to change the pure Python version. Since it won't segfault, there is need for this elaborate defense against something that doesn't arise in normal practice.

**Comment by picnixz** [comment]
> Since it won't segfault, there is need for this elaborate defense against something that doesn't arise in normal practice.

Well.. without those changes, the tests would be broken (no exception would be raised actually). And the C implementation says:

> This implementation is necessarily explicitly equivalent to the pure Python
OrderedDict class in `Lib/collections/__init__.py`.

**Comment by rhettinger** [comment]
> Well.. without those changes, the tests would be broke

You can either make the new tests for the C version only, or you can make the C version emulate what the Python version does (no exception).

> no exception would be raised actually

I'm fine with that.  We made that decision long ago and it has worked out fine in practice.

> This implementation is necessarily explicitly equivalent to the pure Python
OrderedDict class in Lib/collections/__init__.py.

Add, "except where it is necessary to protect the C code from a segfault."

**Comment by picnixz** [comment]
> you can make the C version emulate what the Python version does (no exception).

With this approach, this means I need duplicate the items buffer the entire nodes in advance otherwise calling `clear()` in between would free nodes that are no more available... I'm worried about performances but I'll try to benchmark and see the impact.

> Our only goal in this PR is make sure the C version won't segfault. There should be no other user visible API change.

Actually, the Python implementation does segfault (well it's a segfault in the form of an AttributeError but it's essentially the same issue I think). The Python implementation of `__delitem__` does:

```python
    def __delitem__(self, key, dict_delitem=dict.__delitem__):
        'od.__delitem__(y) <==> del od[y]'
        # Deleting an existing item uses self.__map to find the link which gets
        # removed by updating the links in the predecessor and successor nodes.
        dict_delitem(self, key)
        link = self.__map.pop(key)
        link_prev = link.prev
        link_next = link.next
        link_prev.next = link_next
        link_next.prev = link_prev
        link.prev = None  # <- this has side-effects in __reversed__
        link.next = None  # <- this has side-effects in __iter__
```

And in `__eq__` we do:

```python
	return dict.__eq__(self, other) and all(map(_eq, self, other)))
```

The following test:

```python
        count = 0
        class ChangeExternItem:
            def __eq__(self, other):
                nonlocal count
                if count == 1:
                    del lhs[self]
                count += 1
                return True
            def __hash__(self):
                return -1

        l1, r1 = ChangeExternItem(), ChangeExternItem()
        lhs = self.OrderedDict(dict.fromkeys((0, l1)))
        rhs = self.OrderedDict(dict.fromkeys((0, r1)))
		# AttributeError: 'NoneType' object has no attribute 'key'
		lhs == rhs
``` 

Now, the first `dict.__eq__(lhs, rhs)` call would trigger `del lhs[self]` which itself triggers `OrderedDict.__delitem__`, so in `__iter__`, after you're done getting the key of some item (here, `l1`), you would then do `.next`. However this one will be `None` (so it's not `self.__root`) and you would end up doing `yield curr.key` which fails!

So I still think that the Python implementation should be protected the same the C implementation is. I didn't catch the exception in `.clear()` (don't know why though) but if you delete a key during the iteration, then you end up having some error like that.

In summary the Python implementation has also a bug which can be prevented using the current guards.

**Comment by rhettinger** [comment]
> well it's a segfault in the form of an AttributeErro

That is not a segfault. It is a normal way for functions and methods to die when fed bogus inputs.

**Comment by picnixz** [comment]
> I was not able to reproduce this. The code in the referenced issue terminated normally, giving the output:

Yes, that's what I thought as well. Actually, the `.clear()` is not the issue here (probably because how it's actually implemented) but the problem is with the test example I gave with the `del`. 

> That is not as severe as a segfault

Maybe I should have phrased differently in the sense that the reason why such AttributeError arises is the same reason why it segfaults (at least for this specific test case I put).

>  You've explored state tracking (an approach we intentionally decided not to do); now look for less impactful approaches.

I see that I forgot to put that in my previous comment but I've thought of copying the keys before calling `==` but it comes at a high cost which I'd prefer not to do (i,e, making the C version equivalent to `list(self.keys()) == list(other.keys())`). And even if I were to this, I don't know what to return as a result of the comparison. For instance, if calling `==` has a side effect which changes a value that would previously be compared equal, what should I do? in some sense, the keys are equal and beforehand we already verified that the values are equal, but now it's no more the case, so what should `self == other` return in this case? By catching any mutation, at least I don't need to answer that question (or make a choice).

I'll try to come up with other ideas (it's something I didn't thought of, but maybe I could try to see how dicts are compared in general as well and try to apply that logic here, or see how other languages deal with that issue).

File: Objects/odictobject.c [hunk 1]
```diff
@@ -806,30 +807,50 @@ _odict_keys_equal(PyODictObject *a, PyODictObject *b)
...
-    while (1) {
-        if (node_a == NULL && node_b == NULL)
-            /* success: hit the end of each at the same time */
-            return 1;
-        else if (node_a == NULL || node_b == NULL)
-            /* unequal length */
+    while (node_a != NULL && node_b != NULL) {
+        int res = PyObject_RichCompareBool((PyObject *)_odictnode_KEY(node_a),
+                                           (PyObject *)_odictnode_KEY(node_b),
+                                           Py_EQ);
```
Comment (serhiy-storchaka) [inline Objects/odictobject.c]: Bad things can happen if the key is disappear inside `PyObject_RichCompareBool()`. We should hold references to keys:
```c
PyObject *key_a = Py_NewRef(_odictnode_KEY(node_a));
...
Py_DECREF(key_a);
```

File: Objects/odictobject.c [hunk 2]
```diff
@@ -806,30 +807,50 @@ _odict_keys_equal(PyODictObject *a, PyODictObject *b)
...
+        if (res < 0) {
```
Comment (serhiy-storchaka) [inline Objects/odictobject.c]: What if use `res <= 0` here?

File: Objects/odictobject.c [hunk 3]
```diff
@@ -806,30 +807,50 @@ _odict_keys_equal(PyODictObject *a, PyODictObject *b)
...
+            return res;
+        }
+        else if (a->od_state != state_a || b->od_state != state_b) {
+            // If the size changed, then the state must also have changed
+            // since the former changes only if keys are added or removed,
+            // which in turn updates the state.
+            PyErr_SetString(PyExc_RuntimeError,
+                            (size_a != PyODict_SIZE(a) || size_b != PyODict_SIZE(b))
+                                ? "OrderedDict changed size during iteration"
+                                : "OrderedDict mutated during iteration");
```
Comment (serhiy-storchaka) [inline Objects/odictobject.c]: I am not sure that it is so important to distinguish these two cases. What error message is used depends on implementation details, we cannot guarantee anything here. This can be simplified to always use the same error message,

File: Objects/odictobject.c [hunk 4]
```diff
@@ -806,30 +807,50 @@ _odict_keys_equal(PyODictObject *a, PyODictObject *b)
...
-            /* otherwise it must match, so move on to the next one */
-            node_a = _odictnode_NEXT(node_a);
-            node_b = _odictnode_NEXT(node_b);
         }
+
+        /* otherwise it must match, so move on to the next one */
+        node_a = _odictnode_NEXT(node_a);
+        node_b = _odictnode_NEXT(node_b);
     }
+    assert(node_a == NULL && node_b == NULL);
```
Comment (serhiy-storchaka) [inline Objects/odictobject.c]: I am not so sure. If the Python code in `__eq__` calls `move_to_end()`, the result will be wildly unpredicable. It does not matter, we should not guarantee sensible result in this case, the code simply should not crash.

I suggest to remove this assert. You can add check for the case if they not both equal to NULL, as in the current code.

File: Objects/odictobject.c [hunk 5]
```diff
@@ -806,30 +807,50 @@ _odict_keys_equal(PyODictObject *a, PyODictObject *b)
 {
     _ODictNode *node_a, *node_b;
 
+    // keep operands' state and size to detect undesired mutations
+    const Py_ssize_t size_a = PyODict_SIZE(a);
+    const Py_ssize_t size_b = PyODict_SIZE(b);
+    if (size_a == -1 || size_b == -1) {
+        return -1;
+    }
+    if (size_a != size_b) {
```
Comment (serhiy-storchaka) [inline Objects/odictobject.c]: Should not `dict.__eq__` already check this?

**Review by serhiy-storchaka — COMMENTED** [review]
I agree with @rhettinger: no need to change the Python code.

File: Objects/odictobject.c (hunk 2 above)
Comment (picnixz) [reply Objects/odictobject.c]: Errr.... good catch, thanks! Didn't think of that (I'm too used to the `if (... <0)` construction since it's "the path for handling errors").

File: Objects/odictobject.c (hunk 3 above)
Comment (picnixz) [reply Objects/odictobject.c]: Keeping the "mutated" during iteration should be fine then I think (I actually based my implementation on Nikita's). Also, those messages are the same as those in `odictiter_nextkey` so I thought it was fine.

File: Objects/odictobject.c (hunk 1 above)
Comment (picnixz) [reply Objects/odictobject.c]: Wouldn't it be impossible because of: https://github.com/python/cpython/issues/119004#issuecomment-2108428409?

File: Objects/odictobject.c (hunk 5 above)
Comment (picnixz) [reply Objects/odictobject.c]: Yes, but it does not mean that in-between `dict.__eq__` did not cause side-effects (just use a value that mutates its container and it would still pass). The previous version did:

```C
        else if (node_a == NULL || node_b == NULL)
            /* unequal length */
			return 0;
```  

So it was not necessarily the case that the length is equal. However, I will revert this line because it's maybe easier to just leave everything to the while loop instead.

File: Objects/odictobject.c (hunk 4 above)
Comment (picnixz) [reply Objects/odictobject.c]: I forgot about `move_to_end`, so I will follow your advice.

File: Objects/odictobject.c (hunk 2 above)
Comment (picnixz) [reply Objects/odictobject.c]: I remember why I did that. You might have `== 0` but this doesn't mean that the state might not have been affected. So the `if` must come *after* the `if` on the state. But the `if` on the state should come after the `if` of the `==` in case something more critical happened.

File: Objects/odictobject.c (hunk 1 above)
Comment (picnixz) [reply Objects/odictobject.c]: (Ah, maybe you mean as an additional step to use?)

Comment (serhiy-storchaka) [reply Objects/odictobject.c]: Yes, it is not enough, because nodes themself can be freed. But it may be needed for other issue, which is difficult to reproduce (or impossible in Python). So I have no reproducer, it is just my gut says that.

File: Objects/odictobject.c (hunk 2 above)
Comment (serhiy-storchaka) [reply Objects/odictobject.c]: It may be not a problem. The code does not crash, and this is only what is needed in such case.

It does not matter, write what is better for you.

File: Doc/library/collections.rst [hunk 6]
```diff
@@ -1186,7 +1189,11 @@ anywhere a regular dictionary is used.
    method.
 
 .. versionchanged:: 3.9
-    Added merge (``|``) and update (``|=``) operators, specified in :pep:`584`.
+   Added merge (``|``) and update (``|=``) operators, specified in :pep:`584`.
+
+.. versionchanged:: 3.14
```
Comment (picnixz) [inline Doc/library/collections.rst]: Now I'm wondering how to say it here if the Python implementation does not raise such error =/ (it's still important to say that the exception is raised I think)

**Review by picnixz — COMMENTED** [review]
For now, I've left the messages since they were also used in other functions of this module. Changing them at one place does not feel right to me, so I think we should change them simultaneously if needed.

File: Doc/library/collections.rst (hunk 6 above)
Comment (serhiy-storchaka) [reply Doc/library/collections.rst]: I suggest to remove this directive.

Comparing two collections implies iterating, and iterating may raise a RuntimeError if the dict was mutated.

Also, this is a bugfix, and it will be backported.

File: Lib/collections/__init__.py [hunk 7]
```diff
@@ -314,6 +314,13 @@ def __eq__(self, other):
         while comparison to a regular mapping is order-insensitive.
 
         '''
+        # The Python implementation differs from the C implementation in the
+        # sense that it does not track mutations occurring in __eq__() of keys
+        # or values.
+        #
+        # Since it was decided not to change the Python implementation,
+        # calling ``del self[key]`` in the ``key.__class__.__eq__`` may
+        # raise an AttributeError during iteration.
```
Comment (serhiy-storchaka) [inline Lib/collections/__init__.py]: No need to add this comment.

File: Objects/odictobject.c [hunk 8]
```diff
@@ -699,7 +699,7 @@ _odict_add_new_node(PyODictObject *od, PyObject *key, Py_hash_t hash)
 
     _odictnode_KEY(node) = key;
     _odictnode_HASH(node) = hash;
-    _odict_add_tail(od, node);
+    _odict_add_tail(od, node); // this updates 'od_state'
```
Comment (serhiy-storchaka) [inline Objects/odictobject.c]: I do not think this comment is necessary. On one hand, not every call of `_odict_add_tail()` has such comment -- this implies that calls with a comment are special, but they are not. On other hand, if you add it for every call, it will became trivial.

File: Objects/odictobject.c [hunk 9]
```diff
@@ -806,24 +807,46 @@ _odict_keys_equal(PyODictObject *a, PyODictObject *b)
...
-            else if (res == 0)
+            }
+            else if (a->od_state != state_a || b->od_state != state_b) {
+                // If the size changed, then the state must also have changed
+                // since the former changes only if keys are added or removed,
+                // which in turn updates the state.
+                PyErr_SetString(PyExc_RuntimeError,
+                                (size_a != PyODict_SIZE(a) || size_b != PyODict_SIZE(b))
+                                    ? "OrderedDict changed size during iteration"
+                                    : "OrderedDict mutated during iteration");
```
Comment (serhiy-storchaka) [inline Objects/odictobject.c]: Simpler code is enough.
```suggestion
                PyErr_SetString(PyExc_RuntimeError, "OrderedDict mutated during iteration");
```

**Review by serhiy-storchaka — COMMENTED** [review]
This PR looks quite overcomplicated, especially tests. There are 10 tests for the C implementation, and complex helper class with many attributes and methods. Usually only 1-2 tests are added for similar bugfixes. I andrestand that you want to test many different cases, but are they all necessary? Would not test be simpler if you inline the code of `setup()` and `teardown()`?

File: Objects/odictobject.c (hunk 9 above)
Comment (picnixz) [reply Objects/odictobject.c]: Ok! that way I can refactor the tests as well.

File: Objects/odictobject.c (hunk 8 above)
Comment (picnixz) [reply Objects/odictobject.c]: Ah I forgot to remove them. It was for me during debugging and for tracking where the state changes actually.

**Comment by picnixz** [comment]
> Would not test be simpler if you inline the code of setup() and teardown()?

I'll make a commit with this approach and see how it renders. But yes, it might be easier to review. 

> I andrestand that you want to test many different cases, but are they all necessary

I think I can remove the asymmetric cases now (or the symmetric ones maybe).

**Comment by picnixz** [comment]
@rhettinger Friendly ping in case you forgot about this PR

**Review by serhiy-storchaka — APPROVED** [review]
LGTM.

File: Misc/NEWS.d/next/Library/2024-07-03-14-23-04.gh-issue-119004.L5MoUu.rst [hunk 10]
```diff
@@ -0,0 +1,2 @@
+Fix a crash in :ref:`OrderedDict.__eq__ <collections_OrderedDict__eq__>`
+when operands are mutated during the check.  Patch by Bénédikt Tran.
```
Comment (Eclips4) [inline Misc/NEWS.d/next/Library/2024-07-03-14-23-04.gh-issue-119004.L5MoUu.rst]:
```suggestion
when operands are mutated during the check. Patch by Bénédikt Tran.
```

Comment (serhiy-storchaka) [reply Misc/NEWS.d/next/Library/2024-07-03-14-23-04.gh-issue-119004.L5MoUu.rst]: This is not wrong.

Comment (Eclips4) [reply Misc/NEWS.d/next/Library/2024-07-03-14-23-04.gh-issue-119004.L5MoUu.rst]: > This is not wrong.

I've checked other `NEWS` entries and there's a mix of two or one space after the comma. I'm don't know why this is not wrong, so feel free to ignore my comment.

Comment (picnixz) [reply Misc/NEWS.d/next/Library/2024-07-03-14-23-04.gh-issue-119004.L5MoUu.rst]: It's common to use 2 spaces after a period just for reading purposes. Historically, this is because of typewriters where it could be hard for human eyes to spot the end of a sentence. Nowadays, this can be true for some fonts, especially monospace ones where everything has the same width. 

I personally don't use 2-spaces so thanks for spotting it (though I'm not consistent all the time and I'm pretty sure I've committed rst content with 2 spaces after a period).

**Comment by rhettinger** [comment]
At the sprint, I discussed this briefly with Eric Snow (the original author). He concurred that the pure python code should not change, and he had no objections to adding the state counter for the C version.

**Comment by picnixz** [comment]
Hum, do we backport it to 3.12 and 3.13? (AFAICT, this is a bug fix so there should be a backport)
